sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -229,6 +229,31 @@ class AstBuilder extends DataTypeAstBuilder
     )
   }
 
+  override def visitSearchedCaseStatement(ctx: SearchedCaseStatementContext)
+  : SearchedCaseStatement = {
+    SearchedCaseStatement(
+      conditions = ctx.conditions.asScala.toList.map(boolExpr => withOrigin(boolExpr) {
+        SingleStatement(
+          Project(
+            Seq(Alias(expression(boolExpr), "condition")()),
+            OneRowRelation()))
+      }),
+      conditionalBodies = ctx.conditionalBodies.asScala.toList.map(body => visitCompoundBody(body)),
+      elseBody = Option(ctx.elseBody).map(body => visitCompoundBody(body)))
+  }
+
+  override def visitSimpleCaseStatement(ctx: SimpleCaseStatementContext): SearchedCaseStatement = {
```

Comment:
Simple case statements are currently parsed to searched case statements, which is probably suboptimal because the main case expression (caseVariable) will be evaluated multiple times.

Reply:
I'm not sure about this at the moment, let's think about it on Monday and try to figure out a better approach...

Reply:
> probably suboptimal because the main case expression (caseVariable) will be evaluated multiple times.

I think it is ok for the initial implementation. More important to implement it correctly, and test it properly. @dusantism-db You can open a follow up task for future optimizations.